### ISSUE TITLE
Question: persist and accessories 

### ISSUE THREAD

**Issue by mriksman** [nfarina/homebridge#1614]
I know many times there is an issue, we are told to delete `persist `and/or `accessories`. But what exactly _are_ these folders/files? What does `persist `contain, and what does `accessories `contain? What happens when you delete one or the other (in the Home app particularly)

I accidentally made a code change, which set all the object_ids in `homebridge-wink3` to the same ID. This caused a "Cannot add a bridged Accessory with the same UUID" error. I had to delete the `accessories ` folder (which contains `cachedAccessories`) to make it add fresh again.

Sometimes I delete `accessories`, and when I open the Home app, I watch as they all disappear and then reappear. Sometimes they go back to their assigned rooms, but sometimes they go back to 'Default Room'. So what determines where they end up? What is the expected behaviour?

**Comment by drumfreak** [comment]
Great question... and how long does the cache last? Did you find answers?

**Comment by llamafilm** [comment]
It's a good question.  I would also like to learn how these two "caches" work.

**Comment by ebaauw** [comment]
The `persist` directory contains two files: `AccessoryInfo.`_username_`.json` and `IdentifierCache.`_username_`.json` (where _username_ is the `username` from the `bridge` section in config.json, but without the colons.  `AccessoryInfo` contains the encryption and pairing info with HomeKit; `IdentifierCache` maps the UUID of accessories to the `aid`, and the UUID of services and characteristics to the `sid` (see the HAP spec for what this means).  Deleting these files is like factory resetting homebridge: after that, it no longer responds to HomeKit and needs to be re-added.  Only do this as a last resort, and, preferably, only after removing homebridge from HomeKit.

Homebridge places a single JSON file in the `accessories` directory, `cachedAccessories`, but plugins might create files there as well (e.g. homebridge-hue stores the history files there).  Homebridge persists the bridged accessories exposed by dynamic platform plugins in `cachedAccessories`.  On startup, homebridge loads these, before starting its web server.  If you delete this file (after shutting down homebridge), homebridge will start without exposing these accessories, causing HomeKit to delete them (incl. any associations to rooms, groups, scenes, and automations).  If the plugin re-creates the accessories, HomeKit will see them as new accessories, in the default room (or rather; in the room with the homebridge bridge accessory).  With the introduction of the `-R` flag (see https://github.com/nfarina/homebridge/pull/2190), there shouldn't be any reason to delete `cachedAccessories`.

**Comment by llamafilm** [comment]
@ebaauw this makes a lot of sense, thanks.  So if I understand this right, most plugins create dynamic accessories based on the contents of `config.json` and these will be re-created if you delete `cachedAccessories`.  Is there another kind of plugin that makes 'non-dynamic' accessories?

What is the proper way to delete an accessory?  If I delete it in `config.json` then it will be automatically removed from `cachedAccessories` and also removed from HomeKit?

**Comment by ebaauw** [comment]
> most plugins create dynamic accessories 

No, only `platform` plugins that use the newer dynamic platform accessory model.  `platform` plugins using the older static accessory model as well as `accessory` plugins re-create their accessories on each startup, before the homebridge server starts.

> based on the contents of config.json

Typically that's for `accessory` plugins.  `platform` plugins typically discover the devices to expose as accessories.

> What is the proper way to delete an accessory?

Configure the plugin no longer to expose it, or uninstall the plugin.  Only for leftover accessories from dynamic platform plugins you need to run `homebridge -R` or clear `cachedAccessories`.
